src/Beacon.jsx

```diff
@@ -1,111 +1,206 @@
+    this.targetElement = this.getTargetElement();
   }
 
-  componentDidUpdate(prevProps) {
-    if (prevProps.parent !== this.props.parent) {
-      const { parent } = this.props;
-      this.setState({ parentEl: document.querySelector(parent) });
+  componentWillUpdate() {
+    // If margins are similar then center tooltip
+    if (this.targetElement) {
+      const bounds = this.targetElement.getBoundingClientRect();
+      const centerTooltip = Math.abs(window.innerWidth - bounds.right - bounds.left) < TOOLTIP_TOLERANCE;
+      if (this.state.centerTooltip !== centerTooltip) {
+        this.setState({ centerTooltip });
+      }
     }
   }
 
-  getCoordinates(position, bounds) {
-    switch (position) {
-    case 'top':
-      return {
-        left: bounds.left + ((bounds.right - bounds.left - BEACON_WIDTH) / 2),
-        top: bounds.top
-      };
-    case 'bottom':
-      return {
-        left: bounds.left + ((bounds.right - bounds.left - BEACON_WIDTH) / 2),
-        top: bounds.bottom - BEACON_HEIGHT
-      };
-    case 'right':
-      return {
-        left: bounds.right - BEACON_WIDTH,
-        top: bounds.top + ((bounds.bottom - bounds.top - BEACON_HEIGHT) / 2)
-      };
-    default:
-    case 'left':
-      return {
-        left: bounds.left,
-        top: bounds.top + ((bounds.bottom - bounds.top - BEACON_HEIGHT) / 2)
-      };
+  componentDidUpdate() {
+    this.targetElement = this.getTargetElement();
+  }
+
+  getTargetElement() {
+    if (!this.refs.tooltipTarget) {
+      return null;
+    } else if (this.refs.tooltipTarget instanceof React.Component) {
+      return ReactDOM.findDOMNode(this.refs.tooltipTarget);
+    } else {
+      return this.refs.tooltipTarget;
     }
   }
 
+  getTargetClone() {
+    const clone = this.targetElement.cloneNode(true);
+    clone.removeAttribute('id');
+    // Remove `-webkit-background-composite` style since browser complains about it being deprecated
+    const computedStyle = getComputedStyle(this.targetElement).cssText.replace(/-webkit-background-composite: [^;]+;/, '');
+    clone.setAttribute('style', computedStyle);
+    clone.style.position = 'absolute';
+    clone.style.left = `${this.targetElement.getBoundingClientRect().left + window.pageXOffset}px`;
+    clone.style.top = `${this.targetElement.getBoundingClientRect().top + window.pageYOffset}px`;
+    clone.style.margin = '0px';
+    clone.style.zIndex = this.targetElement.zIndex + 1;
+    clone.className = 'tour-clone';
+    setTimeout(() => clone.className = 'tour-clone tour-highlighted', 0);
+    clone.id = TARGET_CLONE_ID;
+    return clone;
+  }
+
+  renderBeacon(persistent) {
+    if (persistent && !this.state.hash) {
+      // We need to wait until the hash is set before we render.
+      // If it is never set, that means that the user has already
+      // clicked on this beacon so we don't display it again.
+      return Children.toArray(this.props.children)[0];
```

Comment:
This is really weird, you treat `this.props.children` here as `node` but three lines below you are treating it as `element`